File: src/main/java/org/mtransit/android/ui/view/common/StickyHeaderItemDecorator.java
package org.mtransit.android.ui.view.common;

import android.graphics.Canvas;
import android.view.View;
import android.view.ViewGroup;
import android.view.ViewTreeObserver;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

public class StickyHeaderItemDecorator<VH extends RecyclerView.ViewHolder> extends RecyclerView.ItemDecoration {

	@NonNull
	private final StickyAdapter<VH> adapter;
	private int currentStickyPosition = RecyclerView.NO_POSITION;
	@NonNull
	private final RecyclerView recyclerView;
	@NonNull
	private final VH currentStickyHolder;
	private View lastViewOverlappedByHeader = null;

	public StickyHeaderItemDecorator(@NonNull StickyAdapter<VH> adapter, @NonNull RecyclerView recyclerView) {
		this.adapter = adapter;
		this.recyclerView = recyclerView;
		currentStickyHolder = adapter.onCreateHeaderViewHolder(recyclerView);
		fixLayoutSize();
		setupCallbacks();

	}

	private void setupCallbacks() {
		recyclerView.addItemDecoration(this);
	}

	@Override
	public void onDrawOver(@NonNull Canvas c, @NonNull RecyclerView parent, @NonNull RecyclerView.State state) {
		super.onDrawOver(c, parent, state);

		RecyclerView.LayoutManager layoutManager = parent.getLayoutManager();
		if (layoutManager == null) {
			return;
		}

		int topChildPosition = RecyclerView.NO_POSITION;
		if (layoutManager instanceof LinearLayoutManager) {
			topChildPosition = ((LinearLayoutManager) layoutManager).findFirstVisibleItemPosition();
		} else {
			View topChild = parent.getChildAt(0);
			if (topChild != null) {
				topChildPosition = parent.getChildAdapterPosition(topChild);
			}
		}

		if (topChildPosition == RecyclerView.NO_POSITION) {
			return;
		}

		View viewOverlappedByHeader = getChildInContact(parent, currentStickyHolder.itemView.getBottom());
		if (viewOverlappedByHeader == null) {
			if (lastViewOverlappedByHeader != null) {
				viewOverlappedByHeader = lastViewOverlappedByHeader;
			} else {
				viewOverlappedByHeader = parent.getChildAt(topChildPosition);
			}
		}
		lastViewOverlappedByHeader = viewOverlappedByHeader;

		int overlappedByHeaderPosition = parent.getChildAdapterPosition(viewOverlappedByHeader);
		int overlappedHeaderPosition;
		int preOverlappedPosition;
		if (overlappedByHeaderPosition > 0) {
			preOverlappedPosition = adapter.getHeaderPositionForItem(overlappedByHeaderPosition - 1);
			overlappedHeaderPosition = adapter.getHeaderPositionForItem(overlappedByHeaderPosition);
		} else {
			preOverlappedPosition = adapter.getHeaderPositionForItem(topChildPosition);
			overlappedHeaderPosition = preOverlappedPosition;
		}

		if (preOverlappedPosition == RecyclerView.NO_POSITION) {
			return;
		}

		if (preOverlappedPosition != overlappedHeaderPosition && shouldMoveHeader(viewOverlappedByHeader)) {
			updateStickyHeader(topChildPosition);
			moveHeader(c, viewOverlappedByHeader);
		} else {
			updateStickyHeader(topChildPosition);
			drawHeader(c);
		}
	}

	private boolean shouldMoveHeader(View viewOverlappedByHeader) {
		int dy = (viewOverlappedByHeader.getTop() - viewOverlappedByHeader.getHeight());
		return (viewOverlappedByHeader.getTop() >= 0 && dy <= 0);
	}

	private void updateStickyHeader(int topChildPosition) {
		int headerPositionForItem = adapter.getHeaderPositionForItem(topChildPosition);
		if (headerPositionForItem != currentStickyPosition && headerPositionForItem != RecyclerView.NO_POSITION) {
			adapter.onBindHeaderViewHolder(currentStickyHolder, headerPositionForItem);
			currentStickyPosition = headerPositionForItem;
		} else if (headerPositionForItem != RecyclerView.NO_POSITION) {
			adapter.onBindHeaderViewHolder(currentStickyHolder, headerPositionForItem);
		}
	}

	private void drawHeader(Canvas c) {
		c.save();
		c.translate(0, 0);
		currentStickyHolder.itemView.draw(c);
		c.restore();
	}

	private void moveHeader(Canvas c, View nextHeader) {
		c.save();
		c.translate(0, nextHeader.getTop() - nextHeader.getHeight());
		currentStickyHolder.itemView.draw(c);
		c.restore();
	}

	private View getChildInContact(RecyclerView parent, int contactPoint) {
		View childInContact = null;
		for (int i = 0; i < parent.getChildCount(); i++) {
			View child = parent.getChildAt(i);
			if (child.getBottom() > contactPoint) {
				if (child.getTop() <= contactPoint) {
					childInContact = child;
					break;
				}
			}
		}
		return childInContact;
	}

	private void fixLayoutSize() {
		recyclerView.getViewTreeObserver().addOnGlobalLayoutListener(new ViewTreeObserver.OnGlobalLayoutListener() {
			@Override
			public void onGlobalLayout() {
				recyclerView.getViewTreeObserver().removeOnGlobalLayoutListener(this);
				int widthSpec = View.MeasureSpec.makeMeasureSpec(recyclerView.getWidth(), View.MeasureSpec.EXACTLY);
				int heightSpec = View.MeasureSpec.makeMeasureSpec(recyclerView.getHeight(), View.MeasureSpec.UNSPECIFIED);

				int childWidthSpec = ViewGroup.getChildMeasureSpec(
						widthSpec,
						recyclerView.getPaddingLeft() + recyclerView.getPaddingRight(),
						currentStickyHolder.itemView.getLayoutParams().width);
				int childHeightSpec = ViewGroup.getChildMeasureSpec(
						heightSpec,
						recyclerView.getPaddingTop() + recyclerView.getPaddingBottom(),
						currentStickyHolder.itemView.getLayoutParams().height);

				currentStickyHolder.itemView.measure(childWidthSpec, childHeightSpec);

				currentStickyHolder.itemView.layout(0, 0,
						currentStickyHolder.itemView.getMeasuredWidth(),
						currentStickyHolder.itemView.getMeasuredHeight());
			}
		});
	}

	public interface StickyAdapter<VH extends RecyclerView.ViewHolder> {

		int getHeaderPositionForItem(int itemPosition);

		void onBindHeaderViewHolder(VH holder, int headerPosition);

		VH onCreateHeaderViewHolder(@NonNull ViewGroup parent);
	}
}

File: app-android/src/main/java/org/mtransit/android/ui/view/common/StickyHeaderItemDecorator.kt
package org.mtransit.android.ui.view.common

import android.graphics.Canvas
import android.view.View
import android.view.ViewGroup
import android.view.ViewTreeObserver.OnGlobalLayoutListener
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.RecyclerView.ItemDecoration
import androidx.core.graphics.withTranslation

class StickyHeaderItemDecorator<VH : RecyclerView.ViewHolder>(
    private val adapter: StickyAdapter<VH>,
    private val recyclerView: RecyclerView,
) : ItemDecoration() {
    private var currentStickyPosition = RecyclerView.NO_POSITION
    private val currentStickyHolder: VH = adapter.onCreateHeaderViewHolder(recyclerView)
    private var lastViewOverlappedByHeader: View? = null

    init {
        fixLayoutSize()
    }

    override fun onDrawOver(c: Canvas, parent: RecyclerView, state: RecyclerView.State) {
        super.onDrawOver(c, parent, state)
        val layoutManager = parent.layoutManager ?: return
        var topChildPosition = RecyclerView.NO_POSITION
        if (layoutManager is LinearLayoutManager) {
            topChildPosition = layoutManager.findFirstVisibleItemPosition()
        } else {
            val topChild = parent.getChildAt(0)
            if (topChild != null) {
                topChildPosition = parent.getChildAdapterPosition(topChild)
            }
        }
        if (topChildPosition == RecyclerView.NO_POSITION) {
            return
        }
        val viewOverlappedByHeader = getChildInContact(parent, currentStickyHolder.itemView.bottom)
            ?: lastViewOverlappedByHeader
            ?: parent.getChildAt(topChildPosition)
            ?: return // !
        lastViewOverlappedByHeader = viewOverlappedByHeader
        val overlappedByHeaderPosition = parent.getChildAdapterPosition(viewOverlappedByHeader)
        val overlappedHeaderPosition: Int
        val preOverlappedPosition: Int
        if (overlappedByHeaderPosition > 0) {
            preOverlappedPosition = adapter.getHeaderPositionForItem(overlappedByHeaderPosition - 1)
            overlappedHeaderPosition = adapter.getHeaderPositionForItem(overlappedByHeaderPosition)
        } else {
            preOverlappedPosition = adapter.getHeaderPositionForItem(topChildPosition)
            overlappedHeaderPosition = preOverlappedPosition
        }
        if (preOverlappedPosition == RecyclerView.NO_POSITION) {
            return
        }
        if (preOverlappedPosition != overlappedHeaderPosition
            && shouldMoveHeader(viewOverlappedByHeader)
        ) {
            updateStickyHeader(topChildPosition)
            moveHeader(c, viewOverlappedByHeader)
        } else {
            updateStickyHeader(topChildPosition)
            drawHeader(c)
        }
    }

    private fun shouldMoveHeader(viewOverlappedByHeader: View): Boolean {
        val dy = viewOverlappedByHeader.top - viewOverlappedByHeader.height
        return viewOverlappedByHeader.top >= 0 && dy <= 0
    }

    private fun updateStickyHeader(topChildPosition: Int) {
        val headerPositionForItem = adapter.getHeaderPositionForItem(topChildPosition)
        if (headerPositionForItem == RecyclerView.NO_POSITION) {
            return
        }
        if (currentStickyPosition != headerPositionForItem) {
            adapter.onBindHeaderViewHolder(currentStickyHolder, headerPositionForItem)
            currentStickyPosition = headerPositionForItem
        }
    }

    private fun drawHeader(c: Canvas) {
        c.withTranslation(0f, 0f) {
            currentStickyHolder.itemView.draw(this)
        }
    }

    private fun moveHeader(c: Canvas, nextHeader: View) {
        c.withTranslation(0f, (nextHeader.top - nextHeader.height).toFloat()) {
            currentStickyHolder.itemView.draw(this)
        }
    }

    private fun getChildInContact(parent: RecyclerView, contactPoint: Int): View? {
        return (0 until parent.childCount).map { index ->
            parent.getChildAt(index)
        }.firstOrNull { child ->
            child.bottom > contactPoint && contactPoint >= child.top
        }
    }

    private fun fixLayoutSize() {
        recyclerView.viewTreeObserver.addOnGlobalLayoutListener(object : OnGlobalLayoutListener {
            override fun onGlobalLayout() {
                recyclerView.viewTreeObserver.removeOnGlobalLayoutListener(this)
                val recyclerViewWidthSpec = View.MeasureSpec.makeMeasureSpec(recyclerView.width, View.MeasureSpec.EXACTLY)
                val recyclerViewHeightSpec = View.MeasureSpec.makeMeasureSpec(recyclerView.height, View.MeasureSpec.UNSPECIFIED)
                val childWidthSpec = ViewGroup.getChildMeasureSpec(
                    recyclerViewWidthSpec,
                    recyclerView.paddingLeft + recyclerView.paddingRight,
                    currentStickyHolder.itemView.layoutParams.width
                )
                val childHeightSpec = ViewGroup.getChildMeasureSpec(
                    recyclerViewHeightSpec,
                    recyclerView.paddingTop + recyclerView.paddingBottom,
                    currentStickyHolder.itemView.layoutParams.height
                )
                currentStickyHolder.itemView.measure(childWidthSpec, childHeightSpec)
                currentStickyHolder.itemView.layout(
                    0, 0,
                    currentStickyHolder.itemView.measuredWidth,
                    currentStickyHolder.itemView.measuredHeight
                )
            }
        })
    }

    interface StickyAdapter<VH : RecyclerView.ViewHolder> {

        fun getHeaderPositionForItem(itemPosition: Int): Int

        fun onBindHeaderViewHolder(holder: VH, headerPosition: Int)

        fun onCreateHeaderViewHolder(parent: ViewGroup): VH
    }
}